app_dart/lib/src/request_handlers/postsubmit_luci_subscription.dart

```diff
@@ -189,8 +187,41 @@ final class PostsubmitLuciSubscription extends SubscriptionHandler {
   //    b) `completed`: update info like status.
   // 2) the task is already completed.
   //    The task may have been marked as completed from test framework via update-task-status API.
-  bool _shouldUpdateTask(bbv2.Build build, firestore.Task task) {
+  bool _shouldUpdateTask(bbv2.Build build, fs.Task task) {
     return build.status != bbv2.Status.SCHEDULED &&
-        !firestore.Task.finishedStatusValues.contains(task.status);
+        !fs.Task.finishedStatusValues.contains(task.status);
+  }
+
+  /// Check if a builder should be rerun.
+  ///
+  /// A rerun happens when a build fails, the retry number hasn't reached the limit, and the build is on TOT.
+  Future<bool> _shouldAutomaticallyRerun(fs.Task task) async {
+    if (!fs.Task.taskFailStatusSet.contains(task.status)) {
+      log.debug('${task.taskName} is not failing. No rerun needed.');
+      return false;
+    }
+    final retries = task.currentAttempt;
+    if (retries > config.maxLuciTaskRetries) {
+      log.info('Max retries reached for ${task.taskName}');
+      return false;
+    }
+
+    final firestoreService = await config.createFirestoreService();
+    final currentCommit = await fs.Commit.fromFirestoreBySha(
+      firestoreService,
+      sha: task.commitSha,
+    );
+    final commitList = await firestoreService.queryRecentCommits(
+      limit: 1,
+      slug: currentCommit.slug,
+      branch: currentCommit.branch,
+    );
+    final latestCommit = commitList.single;
+
+    if (latestCommit.sha != currentCommit.sha) {
+      log.info('Not tip of tree: ${currentCommit.sha}');
+      return false;
+    }
```

Comment:
Can we add a note here that the Merge Queue uses PresubmitLuciSubscription and that this should be safe?

Reply:
Done.